import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32
import random


class SensorNode(Node):
    def __init__(self):
        super().__init__('temp_sensor') #Node oluÅŸturma
  
        self.declare_parameter('publish_frequency', 1.0) #yayÄ±nlama frekans hÄ±zÄ±
        self.declare_parameter('temp_min', 20.0) #minimum sÄ±caklÄ±k
        self.declare_parameter('temp_max', 30.0) #maximum sÄ±caklÄ±k

        self.publisher_ = self.create_publisher(Float32, 'temperature', 10) #Topic oluÅŸturma

        freq = self.get_parameter('publish_frequency').value #oluÅŸturduÄŸumuz deÄŸeri normal bir deÄŸiÅŸkene alÄ±yoruz ki zaman hesaplamasÄ± yapalÄ±m. (T = 1/f)
        self.timer = self.create_timer(1.0 / freq, self.timer_callback) #T = 1/f, zaman iÃ§erisinde Ã§aÄŸÄ±rÄ±lacak fonksiyon

    def timer_callback(self):
        tmin = self.get_parameter('temp_min').value #min sÄ±caklÄ±k normal deÄŸiÅŸkene alÄ±nÄ±r
        tmax = self.get_parameter('temp_max').value #max sÄ±caklÄ±k alÄ±nÄ±r
        temperature = random.uniform(tmin, tmax) #min-max arasÄ± random sayÄ± Ã¼retilir

        msg = Float32() #msg tipi float(32bit)
        msg.data = temperature #datasÄ± Ã¼retilen random Ä±sÄ±ya eÅŸitlenir(listener'da msg.data dediÄŸimizde Ä±sÄ±ya eriÅŸebilmek iÃ§in)
        self.publisher_.publish(msg) #Mesaj basÄ±lÄ±r
        self.get_logger().info(f'ğŸŒ¡ï¸ SÄ±caklÄ±k: {temperature:.2f} Â°C') #Terminalde log olarak gÃ¶rmek iÃ§in log basÄ±lÄ±r
 
def main(args=None):
    rclpy.init(args=args)
    node = SensorNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()